File: golang/longestValidParenthesis.cpp


type Stack []int

class Stack
{
private:
    stack<int> c;

public:
    // empty: check if stack is empty
    bool empty() {
        return this.c.length() == 0
    }

    // push value onto the stack
    void push(int v) {
        this.c.append(v) // Simply append the new value to the end of the stack
    }

    // Remove and return top element of stack. Return false if stack is empty.
    int pop() {
        if ( this.empty() ) {
            return -1; // false
        }
        int tip = this.pop(); // get tip from stack
        return tip;
    }

    // remove tip of stack, replace
    int peek() {
        if ( this.empty() ) {
            return -1; // false
        }
        int tip = this.top(); // get tip from stack
        return tip;
    }
}

int longestValidParentheses(string s) {
    std::stack<int> stack;
    stack.push(-1);
    int maxans = 0;
    // for ( auto idx, ch : s )
    for ( int idx = 0; idx < s.length(); idx++ ) {
        int ch = s[idx];
        switch ( ch ) {
            case '(': {
                stack.push(idx)
                break;
            }
            case ')': {
                if (! stack.empty() ) {
                    stack.pop();
                }
                if ( stack.empty() ) {   
                    stack.push(idx);
                } else {
                    // if (! stack.empty() ) {
                    int tip = stack.top();
                    maxans = max( maxans, (idx - tip));
                    // }
                }
                break;
            }
        }
    }
    return maxans;
}

File: golang/longestValidParenthesis.go

type Stack []int

// IsEmpty: check if stack is empty
func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}

// Push a new value onto the stack
func (s *Stack) Push(v int) {
	*s = append(*s, v) // Simply append the new value to the end of the stack
}

// Remove and return top element of stack. Return false if stack is empty.
func (s *Stack) Pop() (int, bool) {
	if s.IsEmpty() {
		return -1, false
	}
    index := len(*s) - 1 // Get the index of the top most element.
    element := (*s)[index] // Index into the slice and obtain the element.
    *s = (*s)[:index] // Remove it from the stack by slicing it off.
    return element, true
}

// Remove and return top element of stack. Return false if stack is empty.
func (s *Stack) Peek() (int, bool) {
	if s.IsEmpty() {
		return -1, false
	}
    last := len(*s) - 1 // Get the index of the top most element.
    element := (*s)[last] // Index into the slice and obtain the element.
    return element, true
}

func longestValidParentheses(s string) int {
    var maxans int = 0;
    var stack Stack
    stack.Push(-1)
    for idx := 0; idx < len(s); idx++ {
        switch ch := s[idx]; ch {
            case '(':
                stack.Push(idx)
            case ')':
                stack.Pop()
                if stack.IsEmpty() {   
                    stack.Push(idx)
                } else {
                    tip, ok := stack.Peek()
                    if ok && (idx - tip) > maxans { maxans = (idx - tip) }
                }
        }
    }
    return maxans;
}

